core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerClientWithFileAccess.scala

```diff
@@ -0,0 +1,188 @@
+     * @param network name of the network to get the IP address from
+     * @return the ip address of the container
+     */
+    protected def ipAddressFromFile(id: ContainerId, network: String): Future[ContainerIp] = {
+        configFileContents(containerConfigFile(id)).map { json =>
+            val networks = json.fields("NetworkSettings").asJsObject.fields("Networks").asJsObject
+            val specifiedNetwork = networks.fields(network).asJsObject
+            val ipAddr = specifiedNetwork.fields("IPAddress")
+            ContainerIp(ipAddr.convertTo[String])
+        }
+    }
+
+    // See implemented trait for description
+    override def inspectIPAddress(id: ContainerId, network: String)(implicit transid: TransactionId): Future[ContainerIp] = {
+        ipAddressFromFile(id, network).recoverWith {
+            case _ => super.inspectIPAddress(id, network)
+        }
+    }
+
+    // See implemented trait for description
+    def rawContainerLogs(containerId: ContainerId, fromPos: Long): Future[ByteBuffer] = Future {
+        blocking { // Needed due to synchronous file operations
+            var fis: FileInputStream = null
+            try {
+                val file = containerLogFile(containerId)
+                val size = file.length
+
+                fis = new FileInputStream(file)
+                val channel = fis.getChannel().position(fromPos)
+
+                // Buffer allocation may fail if the log file is too large to hold in memory or
+                // too few space is left on the heap, respectively.
+                var remainingBytes = (size - fromPos).toInt
+                val readBuffer = ByteBuffer.allocate(remainingBytes)
+
+                while (remainingBytes > 0) {
+                    val readBytes = channel.read(readBuffer)
+                    if (readBytes > 0) {
+                        remainingBytes -= readBytes
+                    } else if (readBytes < 0) {
+                        remainingBytes = 0
+                    }
+                }
+
+                readBuffer
+            } catch {
+                case e: Exception =>
+                    throw new IOException(s"rawContainerLogs failed on ${containerId}", e)
+
+            } finally {
+                if (fis != null) fis.close()
+            }
+        }
+    }
+}
+
+trait DockerApiWithFileAccess extends DockerApi {
+
+    /**
+     * Obtains the container's stdout and stderr by reading the internal docker log file
+     * for the container. Said file is written by docker's JSON log driver and has
+     * a "well-known" location and name.
+     *
+     * Reads the log file from the specified position to its end. The returned ByteBuffer
+     * indicates how many bytes were actually read from the file.
+     *
+     * For warm containers, the container log file already holds output from
+     * previous activations that have to be skipped. For this reason, a starting position can be specified.
```

Comment:
Drop that line. Separation of concerns-wise, this is not relevant for the client here.